File: calm-quest-c2e0c943/habit_builder/calculate_weights.js
/**
 * Calculate Random Forest feature importance weights
 * This uses a simple approximation based on correlation with stress_level
 * For exact RF weights, run the Python script with pandas/sklearn installed
 */

import { readFileSync, writeFileSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

// Load the CSV file
const csvPath = join(__dirname, '../unified_lifestyle_dataset.csv');
const csvContent = readFileSync(csvPath, 'utf-8');

// Parse CSV
const lines = csvContent.trim().split('\n');
const headers = lines[0].split(',');
const data = lines.slice(1).map(line => {
  const values = line.split(',');
  const row = {};
  headers.forEach((header, i) => {
    row[header.trim()] = parseFloat(values[i]);
  });
  return row;
});

console.log(`Loaded ${data.length} rows`);
console.log(`Features: ${headers.join(', ')}`);

// Calculate correlation with stress_level for each feature
function calculateCorrelation(featureData, stressData) {
  const n = featureData.length;
  const meanFeature = featureData.reduce((a, b) => a + b, 0) / n;
  const meanStress = stressData.reduce((a, b) => a + b, 0) / n;
  
  let numerator = 0;
  let sumSqFeature = 0;
  let sumSqStress = 0;
  
  for (let i = 0; i < n; i++) {
    const diffFeature = featureData[i] - meanFeature;
    const diffStress = stressData[i] - meanStress;
    numerator += diffFeature * diffStress;
    sumSqFeature += diffFeature * diffFeature;
    sumSqStress += diffStress * diffStress;
  }
  
  const denominator = Math.sqrt(sumSqFeature * sumSqStress);
  return denominator === 0 ? 0 : numerator / denominator;
}

// Extract feature columns and stress
const features = ['sleep_hours', 'screen_time_hours', 'exercise_minutes', 'water_intake_liters', 'meditation_minutes'];
const stressData = data.map(row => row.stress_level);

// Calculate absolute correlations (importance proxy)
const correlations = {};
features.forEach(feature => {
  const featureData = data.map(row => row[feature]);
  const corr = Math.abs(calculateCorrelation(featureData, stressData));
  correlations[feature] = corr;
  console.log(`${feature}: correlation = ${corr.toFixed(4)}`);
});

// Normalize to sum to 1 (like RF importance weights)
const totalCorr = Object.values(correlations).reduce((a, b) => a + b, 0);
const weights = {};
Object.keys(correlations).forEach(feature => {
  weights[feature] = correlations[feature] / totalCorr;
});

console.log('\n' + '='.repeat(60));
console.log('Feature Importance Weights (normalized to sum to 1):');
console.log('='.repeat(60));
Object.entries(weights).forEach(([feature, weight]) => {
  console.log(`${feature}: ${weight.toFixed(6)}`);
});

const totalWeight = Object.values(weights).reduce((a, b) => a + b, 0);
console.log(`\nTotal weight sum: ${totalWeight.toFixed(10)} (should be 1.0)`);

// Save to JSON
const outputPath = join(__dirname, 'models/feature_weights.json');
writeFileSync(outputPath, JSON.stringify(weights, null, 2), 'utf-8');

console.log(`\n✅ Weights saved to: ${outputPath}`);

// Print JavaScript format
console.log('\nJavaScript format (for server/index.js):');
console.log('const FEATURE_WEIGHTS = {');
Object.entries(weights).forEach(([feature, weight]) => {
  console.log(`  ${feature}: ${weight.toFixed(6)},`);
});
console.log('};');

File: calm-quest-c2e0c943/habit_builder/calculate_weights.py
import pandas as pd
import numpy as np
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split

# Load the unified dataset
df = pd.read_csv('../unified_lifestyle_dataset.csv')

# Prepare features and target
X = df[[
    "sleep_hours",
    "screen_time_hours",
    "exercise_minutes",
    "water_intake_liters",
    "meditation_minutes"
]]
y = df["stress_level"]

# Split data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train Random Forest
rf = RandomForestRegressor(random_state=42, n_estimators=100)
rf.fit(X_train, y_train)

# Get feature importance
rf_importance = pd.DataFrame({
    "feature": X.columns,
    "importance": rf.feature_importances_
})

# Normalize to sum to 1
rf_importance["weight"] = rf_importance["importance"] / rf_importance["importance"].sum()
rf_importance = rf_importance.sort_values("weight", ascending=False)

print("Feature Importance Weights:")
print(rf_importance)
print("\nWeight Map (for JavaScript):")
print("{")
for _, row in rf_importance.iterrows():
    print(f'  "{row["feature"]}": {row["weight"]:.6f},')
print("}")

# Save to JSON for easy import
import json
weight_map = dict(zip(rf_importance["feature"], rf_importance["weight"]))
with open('../habit_builder/models/feature_weights.json', 'w') as f:
    json.dump(weight_map, f, indent=2)

print(f"\nWeights saved to habit_builder/models/feature_weights.json")

